include/graphblas/base/blas3.hpp

```diff
@@ -442,6 +442,421 @@ namespace grb {
+	 * \note Invalid descriptors will be ignored.
+	 * 
+	 * \endparblock
+	 */
+	template<
+		Descriptor descr = descriptors::no_operation,
+		class Monoid,
+		typename InputType, typename IOType,
+		typename RIT, typename CIT, typename NIT,
+		Backend backend
+	>
+	RC foldl(
+		IOType &x,
+		const Matrix< InputType, backend, RIT, CIT, NIT > &A,
+		const Monoid &monoid,
+		const typename std::enable_if< 
+			!grb::is_object< IOType >::value &&
+			!grb::is_object< InputType >::value &&
+			grb::is_monoid< Monoid >::value, void
+		>::type * const = nullptr
+	) {
+#ifndef NDEBUG
+		const bool should_not_call_base_scalar_unmasked_matrix_foldl = false;
+		assert( should_not_call_base_scalar_unmasked_matrix_foldl );
+#endif
+		(void) A;
+		(void) x;
+		(void) monoid;
+		return UNSUPPORTED;
+	}
+
+	/**
+	 * Return the lower triangular portion of a matrix, strictly below the k-th 
+	 * diagonal (excluded).
+	 *
+	 * @tparam descr      The descriptor to be used (descriptors::no_operation
+	 * 					  if left unspecified).
+	 * @tparam InputType  The type of the elements in the supplied ALP/GraphBLAS
+	 *                    matrix \a A.
+	 * @tparam OutputType The type of the elements in the supplied ALP/GraphBLAS
+	 *                    matrix \a L.
+	 *
+	 * @param[out] L       The lower triangular portion of \a A, strictly below
+	 * 					   the k-th diagonal.
+	 * @param[in]  A       Any ALP/GraphBLAS matrix.
+	 * @param[in]  k       The diagonal above which to zero out \a A.
+	 * @param[in]  phase   The #grb::Phase in which the primitive is to proceed.
+	 *
+	 * @return grb::SUCCESS  When the call completed successfully.
+	 * @return grb::MISMATCH If the dimensions of \a L and \a A do not match.
+	 *
+ 	 * \parblock
+	 * \par Allowed descriptors
+	 * - transpose_matrix: Consider A^T instead of A.
+	 * - no_casting: If the types of \a L and \a A differ, the primitive
+	 * 				 will fail.
+	 * \endparblock
+	 */
+	template<
+		Descriptor descr = descriptors::no_operation,
+		typename InputType,
+		typename OutputType,
+		typename RIT_L, typename CIT_L, typename NIT_L,
+		typename RIT_A, typename CIT_A, typename NIT_A,
+		Backend implementation
+	>
+	RC tril(
+		Matrix< OutputType, implementation, RIT_L, CIT_L, NIT_L > & L,
+		const Matrix< InputType, implementation, RIT_A, CIT_A, NIT_A > & A,
+		const long int k,
+		const Phase &phase = Phase::EXECUTE,
+		const typename std::enable_if<
+			!grb::is_object< OutputType >::value &&
+			!grb::is_object< InputType >::value &&
+			std::is_convertible< InputType, OutputType >::value
+		>::type * const = nullptr
+	) {
+		(void) L;
+		(void) A;
+		(void) k;
+		(void) phase;
+#ifdef _DEBUG
+		std::cerr << "Selected backend does not implement grb::tril()\n";
+#endif
+#ifndef NDEBUG
+		const bool selected_backend_does_not_support_tril = false;
+		assert( selected_backend_does_not_support_tril );
+#endif
+		const RC ret = grb::clear( L );
+		return ret == SUCCESS ? UNSUPPORTED : ret;
+	}
```

Comment:
I don't think `tril` should be a core primitive though right? (It's an algorithm based on select?)

Reply:
(below tril and triu variants also)